src/Bicep.LangServer/Handlers/BicepCodeActionHandler.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Bicep.Core.CodeAction;
+using Bicep.LanguageServer.CompilationManager;
+using Bicep.LanguageServer.Extensions;
+using Bicep.LanguageServer.Utils;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Document;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Bicep.LanguageServer.Handlers
+{
+    public class BicepCodeActionHandler : CodeActionHandler
+    {
+        private readonly ICompilationManager compilationManager;
+
+        public BicepCodeActionHandler(ICompilationManager compilationManager)
+            : base(CreateCodeActionRegistrationOptions())
+        {
+            this.compilationManager = compilationManager;
+        }
+        
+        public override Task<CommandOrCodeActionContainer> Handle(CodeActionParams request, CancellationToken cancellationToken)
+        {
+            var compilationContext = this.compilationManager.GetCompilation(request.TextDocument.Uri);
+
+            if (compilationContext == null)
+            {
+                return Task.FromResult(new CommandOrCodeActionContainer());
+            }
+
+            var requestStartOffset = PositionHelper.GetOffset(compilationContext.LineStarts, request.Range.Start);
+            var requestEndOffset = request.Range.Start != request.Range.End
+                ? PositionHelper.GetOffset(compilationContext.LineStarts, request.Range.End)
+                : requestStartOffset;
+
+            var quickFixes = compilationContext.Compilation.GetSemanticModel().GetAllDiagnostics()
+                .Where(fixable =>
+                    fixable.Span.ContainsInclusive(requestStartOffset) ||
+                    fixable.Span.ContainsInclusive(requestEndOffset) ||
+                    (requestStartOffset <= fixable.Span.Position && fixable.Span.Position + fixable.Span.Length <= requestEndOffset))
+                .OfType<IFixable>()

```

Comment:
Would doing `OfType<IFixable>` first before the offset check eliminate more diagnostics upfront?

Reply:
Good question! I thought about this before and my conclusion is that it depends, but the current implementation should win most of the time.

Supposing that there are `n` diagnostics in total, `m` fixable diagnostics and `k` diagnostics hitting the `span`, filtering on span first will cost `O(n + k)`, and doing `OfType<IFixable>` is `O(n + m)`. If `m < k` then calling `OfType<IFixable>` first is faster, otherwise filtering on span first is better. I would assume that in most cases `k` is `1` and `m >> k` (say we implement a linter in the future, and there are many rules that can produce fixable diagnostics).